src/gui/vector/qgsattributesformmodel.h

```diff
@@ -0,0 +1,459 @@
+        int mHorizontalStretch = 0;
+        int mVerticalStretch = 0;
+        QgsOptionalExpression mVisibilityExpression;
+        RelationEditorConfiguration mRelationEditorConfiguration;
+        QmlElementEditorConfiguration mQmlElementEditorConfiguration;
+        HtmlElementEditorConfiguration mHtmlElementEditorConfiguration;
+        TextElementEditorConfiguration mTextElementEditorConfiguration;
+        SpacerElementEditorConfiguration mSpacerElementEditorConfiguration;
+        QColor mBackgroundColor;
+        bool mCollapsed = false;
+        QgsOptionalExpression mCollapsedExpression;
+        QgsAttributeEditorElement::LabelStyle mLabelStyle;
+    };
+};
+
+
+class AttributeFormTreeItem
+{
+  public:
+    AttributeFormTreeItem() = default;
+    explicit AttributeFormTreeItem( QgsAttributeFormTreeData::AttributeFormTreeItemType itemType, const QString &name, const QString &displayName = QString(), AttributeFormTreeItem *parent = nullptr );
+    explicit AttributeFormTreeItem( QgsAttributeFormTreeData::AttributeFormTreeItemType itemType, const QgsAttributeFormTreeData::DnDTreeItemData &data, const QString &name, const QString &displayName = QString(), AttributeFormTreeItem *parent = nullptr );
+
+    AttributeFormTreeItem *child( int number );
+    AttributeFormTreeItem *firstChild( const QString &name ) const;
+    int childCount() const;
+    //bool insertChildren(int position, int count, int columns);
+    //bool insertColumns(int position, int columns);
+    AttributeFormTreeItem *parent() { return mParent; }
+    //bool removeChildren(int position, int count);
+    //bool removeColumns(int position, int columns);
+    int row() const;
+    QVariant data( int role ) const;
+    bool setData( int role, const QVariant &value );
+
+    /**
+     * Adds a \a child to this item. Takes ownership of the child.
+     */
+    void addChildItem( std::unique_ptr< AttributeFormTreeItem > &&child );
+
+    /**
+     * Deletes all child items from this item.
+     */
+    void deleteChildren();
+
+    QgsAttributeFormTreeData::AttributeFormTreeItemType type() const { return mItemType; }
+
+    QString name() const { return mName; }
+    void setName( const QString &name ) { mName = name; }
+
+    QString displayName() const { return mDisplayName; }
+    void setDisplayName( const QString &displayName ) { mDisplayName = displayName; }
+
+    QIcon icon() const { return mIcon; }
+    void setIcon( const QIcon &icon ) { mIcon = icon; }
+
+  private:
+    QString mName;
+    QString mDisplayName;
+    QIcon mIcon;
+
+    std::vector< std::unique_ptr< AttributeFormTreeItem > > mChildren;
+    QgsAttributeFormTreeData::AttributeFormTreeItemType mItemType;
+
+    QgsAttributeFormTreeData::DnDTreeItemData mItemData;
+    QgsAttributeFormTreeData::FieldConfig mFieldConfigData;
+    QString mFieldNameData;
+
+    AttributeFormTreeItem *mParent;
+};
+
+
+class QgsAttributesAvailableWidgetsModel : public QAbstractItemModel
```

Comment:
I can imagine we'll be able to have an abstract class inheriting from `QAbstractItemModel` and two subclasses of that abstract class, one for `Available Widgets` and one for `Form Layout`. However, I'm not there yet.